### PR TITLE
PersistentVolume & PersistentVolumeClaim API types

### PR DESCRIPTION
**Pull request by markturansky**
This PR is just the API types<sup>*</sup> coming out of #5105.

@thockin @bgrant0607 

<sup>*</sup> sans PersistentVolumeClaimVolumeSource

### CODE REVIEW

File: pkg/api/types.go [hunk 1]
```diff
@@ -195,6 +195,119 @@ type VolumeSource struct {
 	NFS *NFSVolumeSource `json:"nfs"`
 }
 
+// Similar to VolumeSource but meant for the administrator who creates PVs.
+// Exactly one of its members must be set.
+type PersistentVolumeSource struct {
+	// GCEPersistentDisk represents a GCE Disk resource that is attached to a
+	// kubelet's host machine and then exposed to the pod.
+	GCEPersistentDisk *GCEPersistentDiskVolumeSource `json:"persistentDisk"`
```
Comment (thockin) [inline pkg/api/types.go]: rename the JSON to "gcePersistentDisk" to match v1b3

File: pkg/api/types.go [hunk 2]
```diff
@@ -195,6 +195,119 @@ type VolumeSource struct {
...
+	Phase PersistentVolumePhase `json:"phase,omitempty"`
+}
+
+type PersistentVolumeList struct {
+	TypeMeta `json:",inline"`
+	ListMeta `json:"metadata,omitempty"`
+	Items    []PersistentVolume `json:"items,omitempty"`
+}
+
+// a PersistentVolumeClaim is a user's request for and claim to a persistent volume
```
Comment (thockin) [inline pkg/api/types.go]: nit: godocs style wants the type name as the first word,.

File: pkg/api/types.go (hunk 1 above)
Comment (markturansky) [inline pkg/api/types.go]: changed for both VS and PVS.

File: pkg/api/types.go (hunk 2 above)
Comment (markturansky) [inline pkg/api/types.go]: There were two of these.  Both fixed.

**Comment by thockin** [comment]
LGTM.  Integration is broken right now, will need to kick it once a fix is in.

**Comment by markturansky** [comment]
Rebased and squashed.

File: pkg/api/types.go [hunk 3]
```diff
@@ -195,6 +195,119 @@ type VolumeSource struct {
...
+	ListMeta `json:"metadata,omitempty"`
+	Items    []PersistentVolume `json:"items,omitempty"`
+}
+
+// PersistentVolumeClaim is a user's request for and claim to a persistent volume
+type PersistentVolumeClaim struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the volume
```
Comment (quinton-hoole) [inline pkg/api/types.go]: It defines the claim, right?  Perhaps this is a cut 'n paste oversight?

File: pkg/api/types.go [hunk 4]
```diff
@@ -195,6 +195,119 @@ type VolumeSource struct {
...
+	Spec PersistentVolumeClaimSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about a persistent volume.
```
Comment (quinton-hoole) [inline pkg/api/types.go]: Comment as above.  This refers to the claim, not the volume, right?

Comment (markturansky) [inline pkg/api/types.go]: yes, the Claim Spec is how the pod author requests resources.  The Claim Spec is matched against existing PV Specs (containing actuals).   Claim.Status (what the claimant got) will match Volume.Spec (what the volume has).

File: pkg/api/types.go [hunk 5]
```diff
@@ -195,6 +195,119 @@ type VolumeSource struct {
...
+)
+
+type PersistentVolumePhase string
+
+const (
+	// used for PersistentVolumes that are not yet bound
+	VolumeAvailable PersistentVolumePhase = "Available"
+	// used for PersistentVolumes that are bound
+	VolumeBound PersistentVolumePhase = "Bound"
+	// used for PersistentVolumes where the bound PersistentVolumeClaim was deleted
```
Comment (quinton-hoole) [inline pkg/api/types.go]: Suggest adding a comment to the effect that the Released volume needs to be recycled before becoming Available again.

File: pkg/api/types.go (hunk 4 above)
Comment (thockin) [inline pkg/api/types.go]: I think the point was that the comment says "represents the ... volume"

File: pkg/api/types.go (hunk 5 above)
Comment (quinton-hoole) [inline pkg/api/types.go]: As an aside, and probably out of scope for 1.0, do we accommodate the use case where one user writes to a disk and then hands it over to another user to read?  Seems useful, so ideally don't preclude it as a future possibility.

Comment (markturansky) [inline pkg/api/types.go]: Your additional comment suggestion was added to #5800.

Regarding your second comment, not at the same time (can't be ReadWriteOnce and ReadOnlyMany at the same time), but the claim to your volume lives longer than your pod.  You can use ReadWriteOnce to write data to the volume, delete your pod, then mount the volume ReadOnlyMany (assuming a GCE disk) in a ReplController across many pods.

File: pkg/api/types.go [hunk 6]
```diff
@@ -195,6 +195,119 @@ type VolumeSource struct {
 	NFS *NFSVolumeSource `json:"nfs"`
 }
 
+// Similar to VolumeSource but meant for the administrator who creates PVs.
+// Exactly one of its members must be set.
+type PersistentVolumeSource struct {
```
Comment (quinton-hoole) [inline pkg/api/types.go]: I don't understand why we need both this and VolumeSource.  Can't we just re-use PersistentVolumeSource?
Along the same lines, I think that the distinction between a user and an administrator here is somewhat artificial and temporary, and should not be unnecessarily exposed via the API.

In future I imagine that both users and administrators will be able to create PD's, and should both use the same API.

File: pkg/api/types.go [hunk 7]
```diff
@@ -195,6 +195,119 @@ type VolumeSource struct {
...
+	// This is useful for development and testing only.
+	// on-host storage is not supported in any way
+	HostPath *HostPathVolumeSource `json:"hostPath"`
+}
+
+type PersistentVolume struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	//Spec defines a persistent volume owned by the cluster
```
Comment (quinton-hoole) [inline pkg/api/types.go]: As mentioned elsewhere, I'm not convinced that the concept "owned by the cluster" is necessary.  I think that PersistentVolumes are each simply created within a namespace, and "claimed", potentially from a different namespace.

File: pkg/api/types.go (hunk 6 above)
Comment (markturansky) [inline pkg/api/types.go]: @thockin @bgrant0607 suggested this distinction.  Not all volumes are persistent, only a subset.  Volumes may  overlap between the two.  Kubelet has to be made to work with both, which is what Tim was prepping for in #5642.

I tend to agree with you because the implementation is simpler, but then someone just has to ask why an EmptyDir can be a persistent volume (just as I'm using HostPath as a persistent volume for development).

File: pkg/api/types.go (hunk 7 above)
Comment (markturansky) [inline pkg/api/types.go]: I would not mind a discussion around this.  There is nothing that requires PV to have no namespace except that we've been thinking of them as analogs to nodes.

It would be entirely possible for an admin to provision a PV _in a specific namespace_.  Maybe you get the fancy 1TB volume for your data warehouse probject.  Then again, this was thought to be achievable using @derekwaynecarr's quota framework.

File: pkg/api/types.go (hunk 4 above)
Comment (markturansky) [inline pkg/api/types.go]: gotcha.  Both comments in PersistentVolumeClaim are updated.

Comment (quinton-hoole) [inline pkg/api/types.go]: @thockin: Yes.

This is perhaps a good time to add that I find the uses of "Spec" and "Status" somewhat confusing here (although it's entirely possible that this is due to my understanding).

In the general case, I understand "Spec" to mean "what the user asked for", and "Status" to mean "what the system is currently providing w.r.t. the Spec".

I don't see the above principle applied consistently here (either because the above principle is wrong, and I'm confused, or because we need to apply it more consistently in this code).

Comment (quinton-hoole) [inline pkg/api/types.go]: For example, your comment "PV Specs (containing actuals)" illustrates my confusion.

File: pkg/api/types.go (hunk 5 above)
Comment (quinton-hoole) [inline pkg/api/types.go]: OK, gotcha.  So the state machine looks something like:

(initial) -> Available
Available -> (claim)-> Bound
Bound -> (unclaim) -> Released
Released -> (claim) -> Bound
Released -> (recycle) -> Available

In particular the following transition is possible:

Released ->(claim)-> Bound

File: pkg/api/types.go [hunk 8]
```diff
@@ -195,6 +195,120 @@ type VolumeSource struct {
...
+)
+
+type PersistentVolumeClaimPhase string
+
+const (
+	// used for PersistentVolumeClaims that are not yet bound
+	ClaimPending PersistentVolumeClaimPhase = "Pending"
+	// used for PersistentVolumeClaims that are bound
+	ClaimBound PersistentVolumeClaimPhase = "Bound"
+)
```
Comment (quinton-hoole) [inline pkg/api/types.go]: What phase are claims in, between the time when the user requests unbinding, and when they're actually unbound?  e.g. pod termination has been requested, but not yet completed (e.g. due to a network partition).  This gets back to the storage "fencing" discussion that happened at the last contributor summit.

File: pkg/api/types.go (hunk 6 above)
Comment (markturansky) [inline pkg/api/types.go]: Re: the EmptyDir question, with a little code, any volume can likely be persistent.  Otherwise, wouldn't the admin only provision volumes that make sense as persistent volumes?

In other words, a volume is a volume is a volume.  The PV framework around it gives a volume longevity outside a pod, regardless of type of volume.  Splitting them between VS and PVS involves more code and complexity that isn't strictly needed, but I'm not as knowledgeable on the architecture or Go-isms of the project.

File: pkg/api/types.go (hunk 5 above)
Comment (markturansky) [inline pkg/api/types.go]: I've been saying "release a claim" to mean the user deletes their PersistentVolumeClaim object, which unbinds from PV, and allows recycling of the volume.

Deleting your pod doesn't release a volume.  The volume is unmounted from the host.

You still hold your claim check and can use that volume again in another pod.

File: pkg/api/types.go (hunk 8 above)
Comment (markturansky) [inline pkg/api/types.go]: I think claims are only Pending (not yet matched to a backend resource) or Bound.

You can have a bound claim and never use the volume in a pod.

**Comment by markturansky** [comment]
Rebased and green.

File: pkg/api/types.go (hunk 5 above)
Comment (thockin) [inline pkg/api/types.go]: I think the transition of:
    Bound -> (unclaim) -> Released
is a fatal transition.  Once released, the only next-change is to Available (via recycling)

If you want to "hand off" data you hand off the claim.

File: pkg/api/types.go (hunk 6 above)
Comment (thockin) [inline pkg/api/types.go]: Some of the VolumeSource subtypes do not make sense as PersistentVolumeSources.  Maybe they will make sense eventually, but they don't now.  We had a decision to make.  We could either document this for each VolumeSource field and throw errors if used wrongly or we could split the concepts with a promise to merge them if it makes sense later.

Brian and I both came to the latter choice independently, but neither of us is doing the implementation.  It's entirely possible that we're both wrong (again, because neither of us is doing the implementation).  If you feel that way, I'm open to discussing it, but I really want to get something committed here - done is better than perfect.

We can always revisit this with very little consequence, I think (the API will be compatible).

So I'll put it back on you.  I am sympathetic to the argument that the duplication to support two structs and two sets of plugins is cumbersome.  Is it worth spending time on right now?  That's not a leading question - I am not the one tasked with getting this work committed :)

File: pkg/api/types.go (hunk 5 above)
Comment (quinton-hoole) [inline pkg/api/types.go]: Mark, how is the distinction between an unmounted and a released volume represented in the system?  In particular, what is the series of API calls necessary in order to locate an unmounted volume (e.g. if a pod dies or is terminated), and mount it to a new (or existing? pod)?  I assume that replication controllers will do this all automatically?

File: pkg/api/types.go (hunk 7 above)
Comment (thockin) [inline pkg/api/types.go]: There's a general sense that you don't cross namespace boundaries, pretty much ever.  So attaching PVs to a namespace really limits their use (or we break that rule).

There's no concept of a type that exists sometimes in namespaces and sometimes not.  I don't think we want that, but I guess I am open to arguments.  I think it will make @derekwaynecarr cry.

We will eventually have ACLs such that your fancy 1TB volume (just large enough for acceptance testing :) can only be bound to claims by users in your warehouse project.

In short, I think the congruence of this and nodes is powerful and breaking that should be considered precedent setting.  I.e. if we do this, we should consider the same change for nodes.

**Comment by thockin** [comment]
not green

**Comment by markturansky** [comment]
It was green.  I rebased again and now I'm waiting.  No insight into Shippable for me.

File: pkg/api/types.go (hunk 6 above)
Comment (quinton-hoole) [inline pkg/api/types.go]: That makes sense.  I just thought I'd note the apparent duplication.  Happy go with whatever the most expedient solution is for now, as it doesn't seem to restrict us much in future.

**Comment by markturansky** [comment]
Travis came back green.

File: pkg/api/types.go (hunk 5 above)
Comment (markturansky) [inline pkg/api/types.go]: @quinton-hoole 

1.  With the introduce of a /status subresource (suggested back in #5105), Kubelet could update the volume's status to reflect it's newly mounted state.   Until that is added, there's no way for the user to know if their claim's backing volume is successfully mounted anywhere (perhaps via a cloud API?  Wouldn't help some of the other volumes like NFS).

2.  Nothing magical needs to happen for the volume to follow the pod.  Kubelet unmounts orphaned volumes, which would include detaching cloud volumes.   Kubelet on another host mounts the volume if the pod is recreated on that host after a crash of the original host.

File: pkg/api/types.go (hunk 7 above)
Comment (quinton-hoole) [inline pkg/api/types.go]: OK, I understand.  Now is probably not the right time for this discussion, and I'm happy to defer it until aver v1.0, but I think that there's a lot of value in accommodating users sharing things with each other. And I think that whether a user wants to share something with someone else is an authorization concern, and orthogonal to what namespace the thing is in (along the lines of a file system, where things under my home directory ("namespace"), can be made readable to other users).  I suspect that the system is not set up this way today, in which case we should defer this discussion.

File: pkg/api/types.go (hunk 6 above)
Comment (markturansky) [inline pkg/api/types.go]: This separation in the API allows us to keep some volumes from users (assuming the PV API is restricted to admins) and the implementation in #5398 was pretty easy in retrospect -- so long as it's TA'd (Thockin Approved)<sup>&trade;</sup>

File: pkg/api/types.go (hunk 7 above)
Comment (quinton-hoole) [inline pkg/api/types.go]: PS: I share your distaste for types that are sometimes in a namespace and sometimes not. In fact I would be a fan of everything always being in a namespace.  I believe that @lavalamp also belongs to this school of thought.  I will cry together with @derekwaynecarr if the namespace situation gets even more confusing than it already is :-)

Comment (quinton-hoole) [inline pkg/api/types.go]: PPS:  Regarding nodes and volumes being analogs of each other, I can't help thinking that they're fundamentally different in that nodes seem to be in the domain of the system infrastructure, whereas volumes seem to be in the domain of the user application.  But that may be a view created in my head by previous work building IAAS, where cloud providers provide and manage nodes, whereas users create, use and delete volumes.

Comment (markturansky) [inline pkg/api/types.go]: @quinton-hoole Allow me to split this hair for you :)

You are correct that nodes are system infrastructure.  They represent resources in the infrastructure.  A PersistentVolume is the same thing.  It is a provisioned resource.

A PersistentVolumeClaim is the user application domain you refer to.  Users request a slice of resources by creating a claim (as they request a slice of computing resource by creating a pod).  Users create, use, and delete claims.

We are introducing another volume plugin that allows the user to use their claim as their volume.

Comment (quinton-hoole) [inline pkg/api/types.go]: Thanks for so clearly describing your mental model.  It illustrates very well where I think we differ.  Perhaps best discussed in person some time, but I'll dump my brain here in the mean time :-)

In your current design, as I understand it, a PersistentVolumeClaim lays claim to an already existing, mountable volume, which I agree is in the user application domain.  On the other extreme we have what I will call storage pools, for example the EBS service, the GCE PD service, or an on-premise farm of NetApp filers, all of which have API's for creating mountable volumes of specific sizes, performance characteristics etc. Those storage pools are in the infrastructure domain - they are built, capacity planned, maintained etc by cloud providers (or storage administrators in the case of the on-premise NetApp filer farm) and provide the large slicable resource pool to which you refer.  In between these two (volume claims and storage pools) we have the world where the actual volumes are created. I believe this to be logically in the application domain - users should be able to create and delete them on demand.  Right now I think that we have administrators doing that, but I see that as a temporary measure.  It would be worth getting consensus on the answer to this question - can users (in future) create volumes, or only lay claim to volumes that already exist?

File: pkg/api/types.go (hunk 6 above)
Comment (thockin) [inline pkg/api/types.go]: I'll try to work through related PRs today.  I will not throw roadblocks up if I decide I was wrong - we can come back to it, later.

File: pkg/api/types.go (hunk 7 above)
Comment (thockin) [inline pkg/api/types.go]: In the fullness of time users will (with the blessing of clusteradmin) be able to make a claim and have a volume to back that claim be provisioned dynamically.

In the fullness of time users will (with the blessing of clusteradmin) be able to request a pod and have a node to back that pod be provisioned dynamically.

For a rough rule of thumb: if it costs the cluster admin money, user should not be able to do it unless explicitly allowed.  Consider on-prem.  Users are bound by the pool of hardware available - they might be able to auto-provision new hardware, but it involves the clusteradmin interm going to Fry's.

**Comment by thockin** [comment]
shippable is so slow it has not even ACKed this PR yet

**Comment by thockin** [comment]
Try making a no-op change and push -f to see if shippable will pick it up

**Comment by markturansky** [comment]
Rebased and kicked.

**Comment by markturansky** [comment]
Rebased to get #5398.  Waiting on green.

**Comment by markturansky** [comment]
Rebased this morning and Travis came back green.

**Comment by markturansky** [comment]
thanks @thockin you get a short break while I update the client code with Derek's suggestions.  that PR can get reviewed and merged by derek/clayton.  I'll be back to you with the volume manager and claim plugin soon thereafter.
